proposal: Host IP pool with NNCP (#1088)
nmstate/kubernetes-nmstate: design-proposals/host-ip-pool.md

# Host IP pooling with NNCPs.

## Summary

> Provide a mechanism to assign static IP configuration for secondary interfaces on existing and future cluster nodes

Author: Replacing it with "scale up" and "scale down".

## Motivation

Using DHCP as the IP assignment mechanism has some drawbacks:
- Introduce latencies on node restart.
- Has problematic behaviours when the DHCP server is misconfigured.
- It depends on network setup that can be complex.

To overcome the above-mentioned drawbacks, statically configured interfaces taking the address from a
dynamic host IP pool is a solution. Since the interfaces are static there is no
reconfiguration after some time or at node restart.

This functionality was requested on U/S via https://github.com/nmstate/kubernetes-nmstate/issues/725

Spike using whereabouts https://github.com/nmstate/kubernetes-nmstate/pull/1081

### User Stories

> - As a cluster administrator, I want to configure static IPs on secondary NICs on a pool of nodes without creating one per each node.

Author: The feature is only for day 2 and secondary interfaces, this is a good first starting point, we can exapand it further to transform primary interfaces from DHCP to static or even use this same mechanism out of kubernetes-nmstate, but for now this design document is for secondary interfaces.

> - As a cluster administrator, I want to configure static IPs on secondary NICs on a pool of nodes without creating one per each node.

Reviewer: OK, fair enough - please add then the pod default network to the non-goals section. 

I would also introduction an `objectives` (or goals) section, after the motivation.

My mental model for these things is: 
- motivation
- objectives / goals
- non goals
- challenges (if any :) )
- implementation

- As a cluster administrator, I want scale up the cluster with new nodes getting IPs from the policy.
- As a cluster administrator, I want each of these static IPs to be unique, and taken from a defined range, and freed when unused.

### Goals

- Allow users to use NodeNetworkConfigurationPolicy to configure static cluster-unique IP addresses without hard-coding them.

> - If at all possible, we should pick a solution that could be utilized even in the IPI nmstate API.

Author: For baremetal IPI installation the openshift installer has a api that allow to pass Nmstate configuration that will be rendered into NetworkManager keyfiles. So it's really the IPI installer config networkData API.

https://github.com/openshift/installer/blob/969378dc498336d1ac29baa89f75324a44467d5d/pkg/types/baremetal/platform.go#L43

> - If at all possible, we should pick a solution that could be utilized even in the IPI nmstate API.

Author: Adding it to the doc.

### Non-Goals

- This design does not cover IP address configuration of management NIC on day-0/1

## Proposal

### Workflow Description (assign static address with IPPool)

**cluster admin** is a human user responsible for managing the cluster node
networking.

1. The cluster admin create `NodeNetworkIPPool` resources with IP ranges and a names.

> 2. The cluster admin configure node network interfaces statically but with IP from the `NodeNetworkIPPool` using a

Author: Done

`NodeNetworkConfigurationPolicy` referencing them.

> 3. The cluster admin wait for the `NodeNetworkConfigurationPolicy` to success

Reviewer: I still see the wrong sentence .

> #### Variation (not free IP address at the NodeNetworkIPPool)

Author: Done

networking.

1. The cluster admin changes network interface to remove the refernece to the NodeNetworkIPPool (for example harcoding the IP or switching to DHCP).
2. The cluster admin wait for the NodeNetworkConfigurationPolicy to success so the IPs from that interfaces are free now.

### Workflow Description (scaling up cluster)

**cluster admin** is a human user responsible for managing the cluster node
networking.

1. The cluster admin add a new node to the cluster already configured using IPools
2. The cluster admin wait for the NodeNetworkConfigurationPolicy to success so the new node has took address from the NodeNetworkIPPool

### Workflow Description (scaling down cluster)

**cluster admin** is a human user responsible for managing the cluster node
networking.

1. The cluster admin remove a node form a cluster already configured using `NodeNetworkIPool`
2. The cluster admin wait for the `NodeNetworkConfigurationPolicy` to success so the IPs from removed node are free.

### Workflow Description (manual NodeNetworkIPPool clean up)

**cluster admin** is a human user responsible for managing the cluster node
networking.

> 1. The cluster admin find that some IPs are not used anymore but still allocated

Author: Rewording on a more explicit manner.

> 2. The cluster admin reset `NodeNetworkIPPool` allocated status

Author: That's good point, we can Reconcile on status but we have to be carefull not to enter an infinite loop since the status change to change the ippool will also update the status conditions. 

Maybe an annotation is just fine ? At the end this is an emergency situation and in the future it maybe automatize.

What do you think ?

> 4. The cluster admin wait for `NodeNetworkIPPool` status to success

Reviewer: For what it's worth, this is properly described later on. The user just deletes the indexes from the pool, and the reconcile loop will reconstruct it.

> NOTE: This is a manual garbage collector that is good enough for first implementation it can be converted into automatic in the future.

Author: There is an implemntation history, but it has no dates, I am not sure dates fits on a design document.

> NOTE: This is a manual garbage collector that is good enough for first implementation it can be converted into automatic in the future.

Reviewer: I don't mean dates ; just feature blocks. I guess what you have covers this. Does it make sense to add "automating the garbage collector" as an optional last item on that list ?

> ### API Extensions day 2

Reviewer: I think we also need to see the enactment objects to better understand the proposal. At least I think I do.

Author: Done

This proposal add a new CRD to configure and store IP pools (NodeNetworkIPPool) and some
changes at the NNCPs to reference the ippool.

> Following is an example CR for the cluster wide CRD `NodeNetworkIPPool`:

Author: Maybe it's better to have `ClusterIPPool` ? it's a ip

```yaml
apiVersion: nmstate.io/v1
kind: NodeNetworkIPPool
metadata:
  name: traffic-1
spec:
  range:
    cidr: 10.10.10.0/24
    start: 1
    end: 10

>     exclude: [4, 7]

Author: The whole index thing could be a premature optimization since here we are talking about a maginuted of hundreds so using the the whole IP address even for allocations is not an issue ?

>     exclude: [4, 7]

Reviewer: would will you exclude ranges using indexes ?

status:

>   allocated-offsets: [1, 2, 5, 6, 8]

Reviewer: Don't you need to know who owns each of these offsets ? How can you re-sync your pool ? How will you understand if an address is stale/live if not ?

Author: Added the `NodeNetworkConfigurationEnactment` status API to the doc, now is more clear where is the reference.

Reviewer: Cool, I'll take a look later this week. Thanks.

  conditions:
  - type: Available
    status: True
    reason: ReconciledPool
    message: The pool is up to date
  - type: Failure
    status: False

>     reason: ReconciledPool

Author: This are not events this are conditions, and they denote a final state, a user can wait for a Failure condition and different user can wait for a Available condition:
```bash
kubectl wait --for=condition=Available
kubectl wait --for=condition=Failure
```

In this case we are saying that Failure is false since the pool is reconciled.

>     reason: ReconciledPool

Author: Right, keeping just the Available condition with proper Reason and adding a pair of example with the two errors.

`NodeNetworkIPPool` and `allocated-offsets` will be re-calculated.
In this process the reason will be "ReconcilingPool" and `Available`
marked as false, so users can wait on it.

```yaml
  conditions:
  - type: Available
    status: False
    reason: ReconcilingPool
    message: The pool is being updated
  - type: Failure
    status: False
    reason: ReconcilingPool
```

To assign an IP address from a pool to an interface using NNCP, a special
capture syntax is introduced understood by kubernetes-nmstate
`ippool.[name].allocate` that will trigger the ip allocation mechanism,
it will inject the result on the captured state passed to nmpolicy so it can be
consumed by it, the `ippool.[name].allocate` tag will return a struct with the
foollowing fields `ip` and `prefix-length`.

```yaml
apiVersion: nmstate.io/v1
kind: NodeNetworkConfigurationPolicy
metadata:
  name: static-ip
capture:
    address-from-traffic-1: ippool.traffic-1.allocate
    address-from-traffic-2: ippool.traffic-2.allocate
desiredState:
    interfaces:
    - name: eth1
      type: ethernet
      state: up
      ipv4:
        address:
        - ip: "{{ capture.address-from-traffic-1.ip }}"

>           prefix-length: "{{ capture.address-from-traffic-2.prefix-length }}"

Reviewer: I'm aware this is just an example, but I don't think this example make much sense - i.e. getting the IP address from a pool, but the prefix length from another.

The capture looks good.

        dhcp: false
        enabled: true
    - name: eth2
      type: ethernet
      state: up
      ipv4:
        address:
        - ip: "{{ capture.address-from-traffic-1.ip }}"

>           prefix-length: "{{ capture.address-from-traffic-2.prefix-length }}"

Reviewer: I think your `eth1` should get address / prefix length from pool 1, and `eth2` get address / prefix length from pool 2 .

      desiredState:
        interfaces:
        - name: eth1
          type: ethernet
          state: up
          ipv4:
            address:
            - ip: "{{ capture.address-from-foo.ip }}"
              prefix-length: "{{ capture.address-from-foo.prefix-length }}"
            dhcp: false
            enabled: true
        - name: eth2
          type: ethernet
          state: up
          ipv4:
            address:
            - ip: "{{ capture.address-from-bar.ip }}"
              prefix-length: "{{ capture.address-from-bar.prefix-length }}"
            dhcp: false
            enabled: true
    hosts:
      - name: openshift-master-0
        networkData: [static network configuration for master-0]
      - name: openshift-worker-0
        networkData: [static network configuration for worker-0]
```

The new field `platform.baremetal.networkPolicy` will contain the
content of the day 2 `NodeNetworkConfigurationPolicy` and also the `ipPools`
since there is not going to be multiple `networkPolicy` fields.

### Implementation Details/Notes/Constraints

#### Day 2

Since kubernetes-nmstate has a distributed architecture each of the handler

> pods will calculate an allacatable ip from the `NodeNetworkIPPool` status and

Reviewer: ```suggestion
pods will calculate an allocatable ip from the `NodeNetworkIPPool` status and
```

retry the process on conflict

Handlers will check NNCEs at Reconcile and if the ip has being already assigned
it will re-use that and update `NodeNetworkIPPool` status with it, this way the
`NodeNetworkIPPool` can be reset but IPs are keep.

#### Baremetal IPI

The `image-customize-controller` [1] is a golang controller so it would be quite
easy to import ippool library from kubernetes-nmstate and nmpolicy to render the
`networkPolicy`.

[1] https://github.com/openshift/image-customization-controller/blob/3ad18911d14dde1c23b660f99bb05e12a21c99e4/pkg/ignition/builder.go#L79

### Risks and Mitigations

> Overtime from unexpected failure there are risk of having some IPs allocated

Reviewer: Do you mean downtime ?

Author: Yep, that's it.

> Overtime from unexpected failure there are risk of having some IPs allocated

Reviewer: +1.  I don't think I understand what you're trying to say.

> Overtime from unexpected failure there are risk of having some IPs allocated

Author: Reworded.

but not used making the number of available IPs smaller and reaching the end
sooner than expected, for that reseting the `status.allocated-offsets` should

> reconstruct the pool and bring back those IPs.

Reviewer: Is it too obvious to mention that the smaller the pool, the higher the chance of hitting this condition ? 

Consider adding something like "This risk gets higher for shorter IPPools" .

### Drawbacks

> The baremetal IPI Nmstate do a offline NetworkManager keyfile generation so

Reviewer: ```suggestion
The baremetal IPI Nmstate does a offline NetworkManager keyfile generation so
```

there is no network `currentState` that nmpolicy can reference to, so using

> nmpolicy just to reference the ippools may be too overkill.

Reviewer: Might be just me, but I honestly don't understand this one - what does the keyfile generation have to do with the network current state ?

Author: I want to state that nmpolicy is designed to work in conjuntion with `currentState` so snippets from it can be inserted into the `desiredState`. But at baremetal IPI there is no `currentState` so if user try to put a desiredState with captures other than ones related to ippool they will not work.

- IPAM storage mechanism for kubernetes-nmstate use case (day 2)
- Expression "ippool.[name].allocate" feature either at kubernetes-nmstate or nmpolicy they have to call the created IPAM mechanism
- Integrate IPAM, nmpolicy and kubernetes-nmstate for simplest use case, allocate an address for the first time.
- Cover the deallocation scenario where the pool is unreferenced
- Cover the scale up scenario 
- Cover the scale down scenario
- Cover restart scenarios
- Add `networkPolicy` field at baremetal API [openshift-installer API](https://github.com/openshift/installer/blob/master/pkg/types/baremetal/platform.go#L85)
- Glue `networkPolicy` field with IPAM + nmpolicy at [image-customization-controller](https://github.com/openshift/image-customization-controller)
- Test baremetal IPI

## Alternatives

An alternative is to implement directly an IPAM feature at Nmstate with a 
plugin system for ippool retrieval and allocation storage so at day 2 
it will use the `NodeNetworkIPPool` and `NodeNetworkConfigurationEnactment` and 
at baremetal IPI It will read the ippool from the openshift installer config 

> and store the allocations at the nodes file system.

Reviewer: This means each node would be pre-allocated with its own IP range slice, and would manage it, right ? 

Wouldn't that complicate the scale up / down scenarios ? AFAIU, it would impact each node's already existing range.

Author: No, it means that at barmetal IPI the information about allocated IPs per node is store at the filesystem and that this may be converted after starting up into a NNCP that will create the NNCEs but re-using the allocation.

  - Specific to azure
  - Some code has to be put in place to consume it.

### User Stories

- As a cluster administrator, I want to configure static IPs on secondary NICs on a pool of nodes without creating one per each node.
- As a cluster administrator, I want scale up the cluster with new nodes getting IPs from the policy.
- As a cluster administrator, I want each of these static IPs to be unique, and taken from a defined range, and freed when unused.

### Goals

- Allow users to use NodeNetworkConfigurationPolicy to configure static cluster-unique IP addresses without hard-coding them.
- If at all possible, we should pick a solution that could be utilized even in the IPI nmstate API.

### Non-Goals

- This design does not cover IP address configuration of management NIC on day-0/1

## Proposal

**cluster admin** is a user responsible for managing the cluster node
networking.

### Workflow Description (assign static address with IPPool)

1. The cluster admin creates `NodeIPPool` resources with IP ranges and a name.
2. The cluster admin configures node network interfaces statically, with IPs from the `NodeIPPool` using a
`NodeNetworkConfigurationPolicy` referencing them.

> 3. The cluster admin wait for the `NodeNetworkConfigurationPolicy` to success

Reviewer 2: I guess "The cluster admin wait" -> "The cluster admin wait**s**"

#### Variation (exhausted range at the NodeIPPool)

When the IP pool does not have an available IP address, the `NNCP` will fail 

> with a explicit message, so the range has to be

Reviewer: (... and was by now fixed).

> with a explicit message, so the range has to be increased and NNCP applied again.

Reviewer: but why would you want to support an extra flow that is error prone ? If both are error prone, I would strive to support one - not two.

### Workflow Description (removing a reference to NodeIPPool)

1. The cluster admin changes network interface to remove the reference to the NodeIPPool (for example harcoding the IP or switching to DHCP).
2. The cluster admin waits for the NodeNetworkConfigurationPolicy to succeed, freeing the IPs from those interfaces.

### Workflow Description (scaling up cluster)

1.  The cluster admin adds a new node to the already configured cluster using IPools
2. The cluster admin waits for the NodeNetworkConfigurationPolicy to succeed, causing the new node to take the address from the NodeIPPool

### Workflow Description (scaling down cluster)

> 1. The cluster admin removes a node from an already configured cluster using `NodeIPool`

Author: The wording was missleading, after some changes is more clear, it means that the node has a reference to an IP from a `NodeIPPool` so it has to wait for the `NodeNetworkConfigurationPolicy` to success both at scale up and scale down.

    - name: eth2
      type: ethernet
      state: up
      ipv4:
        address:
        - ip: "{{ capture.address-from-traffic-2.ip }}"
          prefix-length: "{{ capture.address-from-traffic-2.prefix-length }}"
        dhcp: false
        enabled: true
```

The allocated offsets will be marked at the `NodeNetworkConfigurationEnactment`
status and will be used to re-construct the nmstate on reboot and to discover
if new allocation if needed or to de-allocate them the pool is no longer 
referenced, this is done on distributed fashion by each handler at their nodes.

```yaml
apiVersion: nmstate.io/v1
kind: NodeNetworkConfigurationEnactment
metadata:
  name: static-ip.node01
status:
  ippool:
    traffic-1: 
      allocated-offsets: [6]
    traffic-2:
      allocated-offsets: [8]
```

### API Extensions baremetal IPI

At baremetal IPI the Nmstate configuration is passed to openshift installer
using the baremetal host `networkConfig` field [1].

[1] https://github.com/openshift/installer/blob/ca8aacd3c931eefabfddafb49cb44b76f24653a6/docs/user/metal/install_ipi.md#describing-hosts

To accommodate the ippools, capture and nmstate new fields will be needed there

> the following is an example of the install-config.yaml from openshift with

Reviewer 2: Wouldn't this then be day-0/1?

Author: Right

status and will be used to re-construct the nmstate on reboot and to discover
if new allocation if needed or to de-allocate them the pool is no longer
referenced, this is done on distributed fashion by each handler at their nodes.

```yaml
apiVersion: nmstate.io/v1
kind: NodeNetworkConfigurationEnactment
metadata:
  name: static-ip.node01
status:
  ippool:
    traffic-1:
      allocated-offsets: [6]
    traffic-2:
      allocated-offsets: [8]
```

### API Extensions baremetal IPI

At baremetal IPI the Nmstate configuration is passed to openshift installer
using the baremetal host `networkConfig` field [1].

[1] https://github.com/openshift/installer/blob/ca8aacd3c931eefabfddafb49cb44b76f24653a6/docs/user/metal/install_ipi.md#describing-hosts

To accommodate the ippools, capture and nmstate new fields will be needed there
the following is an example of the install-config.yaml from openshift with
the new fields:

```yaml
apiVersion: v1
baseDomain: test.metalkube.org
metadata:
  name: ostest
platform:
  baremetal:
    networkPolicy:
      ipPools:
        traffic-1:
          cidr: 10.10.10.0/24
          start: 1
          end: 10

>           exclude: [4, 7]

Reviewer: are these indexes ?... or just the last byte of the IP in decimal format ? Because they look like the latter. 

How will this fit a CIDR with - let's say - 16 bits for the network ?

Author: This example is wrong since `NodeIPPool` was changed since and now start, end and exclude use full IP address, only the allocate-offsets use index since those will be a massive number of entries (we can have cluster with thousand nodes).

2. The cluster admin waits for the `NodeNetworkConfigurationPolicy` to succeed, causing the new node to take the address from the `NodeIPPool`

### Workflow Description (scaling down cluster)

1. The cluster admin removes a node that was taking an address from a `NodeIPPool`.
2. The cluster admin waits for the `NodeNetworkConfigurationPolicy` to succeed, thus freeing the IPs from the removed node.

### Workflow Description (manual NodeIPPool clean up)

1. The cluster admin finds some stale IPs, They are inconsistencies between `NodeIPPool.status` and `NodeNetworkConfigurationEnactment.status`.
2. The cluster admin resets the `NodeIPPool` allocated status specifying an empty list `status.allocated-offsets: []`.
3. The cluster admin waits for `NodeIPPool` status to succeed

NOTE: This is a manual garbage collector that is good enough for first implementation it can be converted into automatic in the future.

### API Extensions day 2

This proposal adds a new CRD to configure and store IP pools (NodeIPPool) and some
changes at the NNCPs to reference the ippool.

Following is an example CR for the cluster wide CRD `NodeIPPool`:

```yaml
apiVersion: nmstate.io/v1

> kind: NodeIPPool

Reviewer: can we see an example with an IPv6 node ip pool for completeness ? 

I'm trying to get my head around the offsets you have defined below and I still don't fully understand them. Are they indexes within the pool or the last bytes of the IP address ? 

If they're the latter, if you have a prefix length < 24, you'll have to encode them in a number, which will make debugging harder.

I would also like to see a scenario for a /16 CIDR, with some exclude ranges in the middle, so I can make sense of them, and get a hold of how hard it will be to debug this when it goes wrong. The idea would be an exercise of mapping the pool's allocated offsets with the IPs in the enactment.

Author: They are indexes more than offsets but they are not the last bytes, I am adding some examples to clarify it.

# Host IP pooling with NNCPs.

## Summary

Provide a mechanism to assign static IP configuration for secondary interfaces
for cluster nodes, supporting cluster scaling up and down.

## Motivation

Using DHCP as the IP assignment mechanism has some drawbacks:
- Introduce latencies on node restart.
- It presents a single point of failure in case the DHCP server is failing
- It depends on network setup that can be complex.

> To overcome the above-mentioned drawbacks, statically configured interfaces taking the address from a

Reviewer 3: The allocation and management of IP address on bare metal machines is a topic the IT industry had to tackle for the early days of computers.

I am sure DHCP and other options exists today which solve this challenge regardless of the kubernetes ecosystem.

I would suggest to examine existing enterprise and telco grade solutions that exist in the market, to make sure we are not re-inventing the wheel.

It is also not that trivial to claim that the DHCP drawbacks described above worth the investment of an in-house IP pool management. I would suggest first interviewing IT experts in this field before we commit to develop such a solution (which from first sight, looks very challenging and will require a lot of investment for the long run).

dynamic host IP pool is a solution. Since the interfaces are static there is no
reconfiguration after some time or at node restart.

This functionality was requested on U/S via https://github.com/nmstate/kubernetes-nmstate/issues/725

Although this proposal will present a solution with an agnostic storage
mechanism, following is a spike done using whereabouts at the storage
https://github.com/nmstate/kubernetes-nmstate/pull/1081

### Similar product

> 

Reviewer 4: @EdDev asks: how does bare-metal IT do this outside the context of k8s?

- Metal3 IPAM: https://metal3.io/blog/2020/07/06/IP_address_manager.html
  - It depends on CAPI (ClusterAPI)
  - It's though for Metal3 baremetal.
- Azure private addresses: https://docs.microsoft.com/en-us/azure/virtual-network/ip-services/private-ip-addresses
  - Specific to azure

>   - Some code has to be put in place to consume it.

Reviewer 4: Can you add a reference to how assisted installer does this on day 1 ?

Reviewer 4: I'm told it's somewhere in https://github.com/openshift/assisted-service/blob/9bae4c3662e4755e80251284aba529318d76c8e9/docs/user-guide/restful-api-guide.md#additional-nmstate-configuration-examples

with a explicit message, the `NodeIPPool` has to be re-created with a bigger
range.

### Workflow Description (removing a reference to NodeIPPool by NNCP re-creation)

1. The cluster admin deletes and creates the NNCP (NNCP with captures
   are not allowed to be modified).
2. The cluster admin waits for the NodeNetworkConfigurationPolicy to succeed.
   the interfaces will have re-allocated IPs.

### Workflow Description (scaling up cluster)

1. The cluster admin adds a new node affected by the `NodeNetworkConfigurationPolicy`.
2. The cluster admin waits for the `NodeNetworkConfigurationPolicy` to succeed, causing the new node to take the address from the `NodeIPPool`

### Workflow Description (scaling down cluster)

1. The cluster admin removes a node that was taking an address from a `NodeIPPool`.
2. The cluster admin waits for the `NodeNetworkConfigurationPolicy` to succeed, thus freeing the IPs from the removed node.

### Workflow Description (manual NodeIPPool clean up)

1. The cluster admin finds some stale IPs, They are inconsistencies between `NodeIPPool.status` and `NodeNetworkConfigurationEnactment.status`.

> 2. The cluster admin resets the `NodeIPPool` following instructions at "API Exntensions/Manual Garbage Collection" chapter.

Reviewer: ```suggestion
2. The cluster admin resets the `NodeIPPool` following instructions at "API Extensions/Manual Garbage Collection" chapter.
```

Reviewer: I would also add a link to that heading.

3. The cluster admin waits for `NodeIPPool` status to succeed

> NOTE: This is a manual garbage collector that is good enough for first implementation it can be converted into automatic in the future.

Reviewer: ```suggestion
**NOTE**: This is a manual garbage collector that is good enough for first implementation it can be converted into automatic in the future.
```

# Host IP pooling with NNCPs.

## Summary

Provide a mechanism to assign static IP configuration for secondary interfaces
for cluster nodes, supporting cluster scaling up and down.

## Motivation

Using DHCP as the IP assignment mechanism has some drawbacks:
- Introduce latencies on node restart.
- It presents a single point of failure in case the DHCP server is failing
- It depends on network setup that can be complex.

To overcome the above-mentioned drawbacks, statically configured interfaces taking the address from a
dynamic host IP pool is a solution. Since the interfaces are static there is no

> reconfiguration after some time or at node restart.

Author: When we hardcode the ip address statically per node, that's there until the user change it, this should work similar to that, we are not implementing DHCP lease.

> reconfiguration after some time or at node restart.

Author: We need the reference to the pool at NNCE so the handler is able to reconstruct the offset at `NodeIPPool` if needed.

> reconfiguration after some time or at node restart.

Reviewer 5: But in case we want the pool to forget the IP, we don't want it to be reconstructed.

> reconfiguration after some time or at node restart.

Author: The user IP will be forget by removing it from NNCE ippools at status, so it will not be bringed back.

> reconfiguration after some time or at node restart.

Author: Is doable technically, but maybe is better to implemnent the mechanism with a annotation. 

```yaml
annotations:
  nmstate.io/reset-ipool: traffic1
```

We  can use this at NNCP (it will all nodes affected by it) or at NNCE (it will affect only the NNCE node).

> reconfiguration after some time or at node restart.

Reviewer: It is a good idea, which I think makes it more API driven, and less prone to errors.

Don't want to get in your way though, you can start small, and implement this after the MVP.

Just for future reference, this is how I see it happening (sure you think the same):
  - user would add the annotation
  - the controller would reconcile the enactments across all the cluster (not getting into detail of if each local agent resets its part of the pool or not)
  - once that's done, the controller removes the annotation.

If this happens, it would be interesting to throw an event / add a condition. 

I would favor the condition since that lives forever, while the event is deleted after a short while.

[1] For baremetal IPI installation the openshift installer has a api that allow
to pass Nmstate configuration that will be rendered into NetworkManager keyfiles.
So it's really the IPI installer config networkData API.
https://github.com/openshift/installer/blob/969378dc498336d1ac29baa89f75324a44467d5d/pkg/types/baremetal/platform.go#L43

### Non-Goals

- This design does not cover IP address configuration of management NIC on day-0/1

## Proposal

**cluster admin** is a user responsible for managing the cluster node
networking.

### Workflow Description (assign static address with IPPool)

1. The cluster admin creates `NodeIPPool` resources with IP ranges and a name.
2. The cluster admin configures node network interfaces statically, with IPs from the `NodeIPPool` using a
`NodeNetworkConfigurationPolicy` referencing them.
3. The cluster admin waits for the `NodeNetworkConfigurationPolicy` to succeed

#### Variation (exhausted range at the NodeIPPool)

When the IP pool does not have an available IP address, the `NNCP` will fail
with a explicit message, the `NodeIPPool` has to increase pool size.

> Modifying `NodeIPPool` means increasing CIDR, changing begin/end or exclude,

Reviewer 5: What if an already allocated IP is not in the range after the change?

Reviewer: +1.

I would also like to see a list of scenarios for which update is possible / not possible.

Keep in mind I think the update scenario is not required, and I have advocated in the past for us to keep the pool's spec immutable.

Author: > What if an already allocated IP is not in the range after the change?

The modification is restricted to not exclude already allocated IPs.

> Keep in mind I think the update scenario is not required, and I have advocated in the past for us to keep the pool's spec immutable.

We can stick with NodeIPPool re-create at this design doc and adding update `NodeIPPool` feature later on. @AlonaKaplan @maiqueb what do you think ?

Reviewer: +1 to keeping the pool spec immutable.

Reviewer 5: When you re-create the pool. What will happen to the nodes pointing to the old deleted pool?

Author: The steps are:
- Create a new NodeIPPool (it has to be in the range of previous one)
- Change NNCPs to reference the new NodeIPPool
- Already reserved IPs will be marked as so in the pool and the node that is missig the IP will get new one.

since allocations are stored as offsets they will have to be re-calculated from
the full IP present at NNCEs.

### Workflow Description (removing a reference to NodeIPPool)

1. The cluster admin changes the network interface to remove the reference to the `NodeIPPool` (for example harcoding the IP or switching to DHCP).
2. The cluster admin waits for the NodeNetworkConfigurationPolicy to succeed, freeing the IPs from those interfaces.

### Workflow Description (scaling up cluster)

1. The cluster admin adds a new node affected by the `NodeNetworkConfigurationPolicy`.
2. The cluster admin waits for the `NodeNetworkConfigurationPolicy` to succeed, causing the new node to take the address from the `NodeIPPool`

> ### Workflow Description (scaling down cluster)

Reviewer 5: What about the flow-  `The cluster admin removes from the node an interface with IP from the `NodeIPPool`?

Author: I will add that workflow

> 1. The cluster admin removes a node that was taking an address from a `NodeIPPool`.

Reviewer 5: Can we have more than one IP on the node from the same `NodeIPPool`?
Can differenct interface on the node take IPs from different `NodeIPPool`s?

Author: > Can we have more than one IP on the node from the same `NodeIPPool`? 
Yep, multiple "captures" will allocate multiple IPs that can be then referenced on the `desiredState`

Can differenct interface on the node take IPs from different `NodeIPPool`s?
The NNCE has a new field at status that show what IPs where taken from what pools, this allow the garbage collector to return the IP to the correct pool.

2. The cluster admin waits for the `NodeNetworkConfigurationPolicy` to succeed, thus freeing the IPs from the removed node.

### Workflow Description (manual NodeIPPool clean up)

1. The cluster admin finds some stale IPs, They are inconsistencies between `NodeIPPool.status` and `NodeNetworkConfigurationEnactment.status`.

> 2. The cluster admin resets the `NodeIPPool` allocated status specifying an empty list `status.allocated-offsets: []`.

Reviewer 5: What is `status.allocated-offsets`? It wasn't mentioned earlier in the doc.

Author: > What is `status.allocated-offsets`? It wasn't mentioned earlier in the doc.

It's later on mentiones on the API section, it's the offset from the "begin" on the range for allocated IPs.

3. The cluster admin waits for `NodeIPPool` status to succeed

NOTE: This is a manual garbage collector that is good enough for first implementation it can be converted into automatic in the future.

### API Extensions day 2

This proposal adds a new CRD to configure and store IP pools (NodeIPPool) and some
changes at the NNCPs to reference the ippool.

Following is an example CR for the cluster wide CRD `NodeIPPool`:

```yaml
apiVersion: nmstate.io/v1

> kind: NodeIPPool

Author: > Are we planning to have the same kind of pool for VMs/pods in the future? If so, the name `NodeIPPool` is a problematic since we won't we able to ruse this crd.

Nope this is just for nodes that's why it has the `Node` preffix.

metadata:
  name: traffic-1
spec:

>   range:

Reviewer: Why would a pool provide addresses from multiple ranges @AlonaKaplan ? 

Assuming there are multiple ranges, If a node requests an address from a pool, from which of the ranges will it get the IP ? The first ? Round-robin ? Random ?...

>   range:

Reviewer 5: yes, randomly one of the ranges.
Regarding the usage, I"m not sure. But for example in the [cisco documentation](https://www.cisco.com/c/en/us/td/docs/net_mgmt/prime/network/4-2-3/user/guide/CiscoPrimeNetwork423UserGuide/ip-pool.pdf) there is a concept of assigning an ip address from multiple ranges (group of pools).

>   range:

Reviewer: I don't know ... Let's say you have a range with ipv6, and another with ipv4. 

This would leave you with either an ipv4 or ipv6 address, depending on... randomness.

If we take this esoteric route (I find it strange) I would at least assure that all ranges belong to the same family.

>   range:

Reviewer: I thought about this a bit more and I honestly don't understand why would a pool feature multiple ranges. It can, but, the same can be done with multiple pools. 

Plus, I think if you have multiple ranges, you'll also have to keep track on those allocations for the different ranges in the pool. I just don't see the motivation for it - thus, I wouldn't account for it **until** there's a use case that requires it.

>   range:

Author: @AlonaKaplan since every range will need it's own allocated info it does not gives any value define multiple of them at the same `NodeIPPool` we are not going to have a massive number of ranges when we talk about interfaces on nodes.

    cidr: 10.10.10.0/24
    start: 10.10.10.1
    end: 10.10.10.10
    exclude: ["10.10.10.4", "10.10.10.7"]
status:
  allocated-offsets: [1, 2, 5, 6, 8]
  conditions:

>   - type: Available

Author: Here https://github.com/nmstate/kubernetes-nmstate/pull/1088/files#diff-f8b6fe96158c42df632deb84deee540aff93c8560a4c4d99fc6db6645797b2efR220-R225

consumed by it, the `ippool.[name].allocate` tag will return a struct with the
following fields `ip` and `prefix-length`.

```yaml
apiVersion: nmstate.io/v1
kind: NodeNetworkConfigurationPolicy
metadata:
  name: static-ip
capture:
    address-from-traffic-1: ippool.traffic-1.allocate
    address-from-traffic-2: ippool.traffic-2.allocate
desiredState:
    interfaces:
    - name: eth1
      type: ethernet
      state: up
      ipv4:
        address:
        - ip: "{{ capture.address-from-traffic-1.ip }}"
          prefix-length: "{{ capture.address-from-traffic-1.prefix-length }}"
        dhcp: false
        enabled: true
    - name: eth2
      type: ethernet
      state: up
      ipv4:
        address:
        - ip: "{{ capture.address-from-traffic-2.ip }}"
          prefix-length: "{{ capture.address-from-traffic-2.prefix-length }}"
        dhcp: false
        enabled: true
```

The allocated IPs will be marked at the `NodeNetworkConfigurationEnactment`
status and will be used to re-construct the nmstate on reboot and to discover
if new allocation if needed or to de-allocate them the pool is no longer
referenced, this is done on distributed fashion by each handler at their nodes.

> The `NNCE.status.ippool` information allows to reconstruct `NodeIPPool.status.allocated-offsets`

Reviewer 5: I still don't get it. If you change the "begin" you should update the offsets, right. But you have the offset list, you know what change was done to the begin, so you can re-calculate the offsets. Why in this case do you need to know what node it using the IP?

> The `NNCE.status.ippool` information allows to reconstruct `NodeIPPool.status.allocated-offsets`

Author: In case we continue allowing `NodeIPPool` update, it's true that offset can be recalculated comparing old a new begin but since this work is going to be distribuged across handlers doing the logic of just registering the IPs they have looks simpler.

when `cidr` and `begin` values are modified.

```yaml
apiVersion: nmstate.io/v1
kind: NodeNetworkConfigurationEnactment
metadata:
  name: static-ip.node01
status:

>   ippool:

Author: Here https://github.com/nmstate/kubernetes-nmstate/pull/1088/files#diff-f8b6fe96158c42df632deb84deee540aff93c8560a4c4d99fc6db6645797b2efR239-R241

pods will calculate an allocatable ip from the `NodeIPPool` status and
retry the process on conflict

Handlers will check NNCEs at Reconcile and if the ip has being already assigned
it will re-use that and update `NodeIPPool` status with it, this way the
`NodeIPPool` can be reset but IPs are kept.

To optimize on storage and lookup instead of using a list of allocated offset
it's possible to use a bitmap, the OVN project use that approach to store the
offsets on a range [1], a snapshot of the bitmap can be stored as base64 at the
`NodeIPPool` status.

[1] https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/ipallocator/allocator.go

#### Baremetal IPI

The `image-customize-controller` [1] is a golang controller so it would be quite
easy to import ippool library from kubernetes-nmstate and nmpolicy to render the
`networkPolicy`.

[1] https://github.com/openshift/image-customization-controller/blob/3ad18911d14dde1c23b660f99bb05e12a21c99e4/pkg/ignition/builder.go#L79

### Risks and Mitigations

If a bug prevents IPs deallocation the ippool can be exhausted and end up
with network reconfiguration or scaling up failures. To fix those situations

> the field `status.allocated-offsets` at `NodeIPPool` can be set with an

Author: > What is the trigger to reconstruct the pool? Empty `status.allocated-offsets`list? 

Yes

> How do we know the pool actually needs a reconstruction and it's not just the the list is empty because no address is used? 

If no address is used the `ippools` field at NNCE will be empty and nothing will be done

> What component is responsible to reconstruct the pool?

Is distributed across the handlers, every handler try to allocate an address for it's interfaces, we do an optimistic locking with retry on failure like we do at NNCP status.

> the field `status.allocated-offsets` at `NodeIPPool` can be set with an

Author: and the NNCP will be marced as not available.

- 252 -> 10.10.12.1
- 253 -> 10.10.12.2
...
- 258 -> 10.10.12.7 (excluded)
- 259 -> 10.10.12.8
- 260 -> 10.10.12.9
- 261 -> 10.10.12.10

IPv4 /116 CIDR pool (4096 IPs)
```yaml
cidr: 2001:4860:4860::/116
start: 2001:4860:4860::000F
end: 2001:4860:4860::00FF
exclude: ["2001:4860:4860::0010", "2001:4860:4860::001F"]
```

offset -> IP:
- 1 -> 2001:4860:4860::000F
- 2 -> 2001:4860:4860::0010 (excluded)
...
- 18 -> 2001:4860:4860::001F (excluded)
- 19 -> 2001:4860:4860::0020
...
- 240 -> 2001:4860:4860::00FF

Reseting `status.allocated-offset` field will trigger a Reconcile on the
`NodeIPPool` and `allocated-offsets` will be re-calculated.
In this process the reason will be "ReconcilingPool" and `Available`
marked as unknown, so users can wait on it.

Following is a table with the "type=Available" condition states

| reason              | status  | message |
| --------------------| --------| ------ |
| ReconcilingPool     | Unknown | The pool is being updated |
| ExhaustedIPPool     | False   | The IP pool has being depleted, manual intervention needed to modify NodeIPPool.|
| UnsyncronizedIPPool | False   | Failed re constructing the IP pool, it may be not in sync with nodes. |

> | OverlapingIPPool    | False   | The NodeIPPool XYZ is overlapping with this one. |

Reviewer 5: We shouldn't get this. But maybe it can be useful to have. `duplicate ip address`, in case two nodes got the same address somehow.

Reviewer 5: I don't think that allocating a new address to one of them is a good option. Both may be running nodes, we cannot just change the IP of one of them.
The tool should prevent getting to this situation, but if the situation somehow happened (due to a bug), it should report about it.

Author: Fair enough, adding this error, we can implement some kind of remediation in the future, this can also come handy if users are configuring statically the cluster and they configured the same ip address at a pair of nodes by mistake.

Author: A user can be already assining the same address statically at the same interface, or at different interfaces, we can detect those.

## Motivation

Using DHCP as the IP assignment mechanism has some drawbacks:
- Introduce latencies on node restart.
- It presents a single point of failure in case the DHCP server is failing
- It depends on network setup that can be complex.

To overcome the above-mentioned drawbacks, statically configured interfaces taking the address from a
dynamic host IP pool is a solution. Since the interfaces are static there is no
reconfiguration after some time or at node restart.

This functionality was requested on U/S via https://github.com/nmstate/kubernetes-nmstate/issues/725

Although this proposal will present a solution with an agnostic storage
mechanism, following is a spike done using whereabouts at the storage
https://github.com/nmstate/kubernetes-nmstate/pull/1081

### Similar product

- Metal3 IPAM: https://metal3.io/blog/2020/07/06/IP_address_manager.html
  - It depends on CAPI (ClusterAPI)
  - It's though for Metal3 baremetal.
- Azure private addresses: https://docs.microsoft.com/en-us/azure/virtual-network/ip-services/private-ip-addresses
  - Specific to azure
  - Some code has to be put in place to consume it.

### User Stories

> - As a cluster administrator, I want to configure static IPs on secondary NICs on a pool of nodes without creating one per each node.

Author: They are usually the NICs tha are not part of control-plane, so if we missconfigure them k8s is still working fine.

> - As a cluster administrator, I want to configure static IPs on secondary NICs on a pool of nodes without creating one per each node.

Reviewer: But how do they look like to the user (cluster-admin) ? I.e. is there a property on the interfaces that can be used to understand which ifaces are secondary ? 

If you don't mind, please define what a "secondary" interface is, to make things clear.

https://github.com/openshift/installer/blob/969378dc498336d1ac29baa89f75324a44467d5d/pkg/types/baremetal/platform.go#L43

### Non-Goals

- This design does not cover IP address configuration of management NIC on day-0/1

## Proposal

**cluster admin** is a user responsible for managing the cluster node
networking.

### Workflow Description (assign static address with IPPool)

1. The cluster admin creates `NodeIPPool` resources with IP ranges and a name.
2. The cluster admin configures node network interfaces statically, with IPs from the `NodeIPPool` using a
`NodeNetworkConfigurationPolicy` referencing them.
3. The cluster admin waits for the `NodeNetworkConfigurationPolicy` to succeed

#### Variation (exhausted range at the NodeIPPool)

When the IP pool does not have an available IP address, the `NNCP` will fail
with a explicit message, the `NodeIPPool` has to increase pool size.

Modifying `NodeIPPool` means increasing CIDR, changing begin/end or exclude,
since allocations are stored as offsets they will have to be re-calculated from
the full IP present at NNCEs.

### Workflow Description (removing a reference to NodeIPPool)

> 1. The cluster admin changes the network interface to remove the reference to the `NodeIPPool` (for example harcoding the IP or switching to DHCP).

Author: Well know that I think of a NNCP with captures cannot be modified, so the only way would be user deleting the NNCP, meaning that all ips are de-allocated.

2. The cluster admin resets the `NodeIPPool` allocated status specifying an empty list `status.allocated-offsets: []`.
3. The cluster admin waits for `NodeIPPool` status to succeed

NOTE: This is a manual garbage collector that is good enough for first implementation it can be converted into automatic in the future.

### API Extensions day 2

This proposal adds a new CRD to configure and store IP pools (NodeIPPool) and some
changes at the NNCPs to reference the ippool.

Following is an example CR for the cluster wide CRD `NodeIPPool`:

```yaml
apiVersion: nmstate.io/v1
kind: NodeIPPool
metadata:
  name: traffic-1
spec:
  range:
    cidr: 10.10.10.0/24
    start: 10.10.10.1
    end: 10.10.10.10
    exclude: ["10.10.10.4", "10.10.10.7"]
status:
  allocated-offsets: [1, 2, 5, 6, 8]
  conditions:
  - type: Available
    status: True
    reason: ReconciledPool
    message: The pool is up to date
apiVersion: nmstate.io/v1
kind: NodeIPPool
metadata:
  name: traffic-2
spec:
  range:
    cidr: 10.10.11.0/24
    start: 10.10.11.1
    end: 10.10.10.11.10
    exclude: ["10.10.11.4", "10.10.11.7"]
status:

>   allocated-offsets: [2, 5, 6, 8]

Author: At the end I am changin the offset to start with "0" since begin + offset should give you the the IP.

```

offset -> IP:
- 1 -> 10.10.11.4
- 2 -> 10.10.11.5 (excluded)
...
- 251 -> 10.10.11.254
- 252 -> 10.10.12.1
- 253 -> 10.10.12.2
...
- 258 -> 10.10.12.7 (excluded)
- 259 -> 10.10.12.8
- 260 -> 10.10.12.9
- 261 -> 10.10.12.10

IPv4 /116 CIDR pool (4096 IPs)
```yaml
cidr: 2001:4860:4860::/116
start: 2001:4860:4860::000F
end: 2001:4860:4860::00FF
exclude: ["2001:4860:4860::0010", "2001:4860:4860::001F"]
```

offset -> IP:
- 1 -> 2001:4860:4860::000F
- 2 -> 2001:4860:4860::0010 (excluded)
...
- 18 -> 2001:4860:4860::001F (excluded)
- 19 -> 2001:4860:4860::0020
...
- 240 -> 2001:4860:4860::00FF

Reseting `status.allocated-offset` field will trigger a Reconcile on the
`NodeIPPool` and `allocated-offsets` will be re-calculated.
In this process the reason will be "ReconcilingPool" and `Available`
marked as unknown, so users can wait on it.

Following is a table with the "type=Available" condition states

| reason              | status  | message |
| --------------------| --------| ------ |
| ReconcilingPool     | Unknown | The pool is being updated |
| ExhaustedIPPool     | False   | The IP pool has being depleted, manual intervention needed to modify NodeIPPool.|

> | UnsyncronizedIPPool | False   | Failed re constructing the IP pool, it may be not in sync with nodes. |

Author: Maybe re-creating NNCPs, I am not sure to be honest

> | OverlapingIPPool    | False   | The NodeIPPool XYZ is overlapping with this one. |

Author: You mean add an admission webhook to check this ?

  name: static-ip
capture:
    address-from-traffic-1: ippool.traffic-1.allocate
    address-from-traffic-2: ippool.traffic-2.allocate
desiredState:
    interfaces:
    - name: eth1
      type: ethernet
      state: up
      ipv4:
        address:
        - ip: "{{ capture.address-from-traffic-1.ip }}"
          prefix-length: "{{ capture.address-from-traffic-1.prefix-length }}"
        dhcp: false
        enabled: true
    - name: eth2
      type: ethernet
      state: up
      ipv4:
        address:
        - ip: "{{ capture.address-from-traffic-2.ip }}"
          prefix-length: "{{ capture.address-from-traffic-2.prefix-length }}"
        dhcp: false
        enabled: true
```

The allocated IPs will be marked at the `NodeNetworkConfigurationEnactment`
status and will be used to re-construct the nmstate on reboot and to discover
if new allocation if needed or to de-allocate them the pool is no longer
referenced, this is done on distributed fashion by each handler at their nodes.

The `NNCE.status.ippool` information allows reserve those IPs again if the
`NodeIPPool` is re-created after exhaustion, since fields on the `NodeIPPool.spec` section
are read-only for simplicity.

> If the IPs reserved for a node for a specific pool want to be deallocated the
> user can change `NNCE.status.ippool` fields if the node is Ready the IP allocation
> will try to get a new one for the desiredState.

Reviewer: This doesn't read well: I don't think IPs want to be deallocated. They don't want anything :P 

```suggestion
The user can change the `NNCE.status.ippool` attribute to deallocate the IPs reserved for a node.

If the node is Ready,  the nmstate will attempt to get a new IP allocation for the desiredState.
```

```yaml
apiVersion: nmstate.io/v1
kind: NodeNetworkConfigurationEnactment
metadata:
  name: static-ip.node01
status:
  desiredState:
    interfaces:
    - name: eth1
      type: ethernet
      state: up
      ipv4:
        address:
        - ip: 10.10.10.8
          prefix-length: 24
        dhcp: false
        enabled: true
    - name: eth2
      type: ethernet
      state: up
      ipv4:
        address:
        - ip: 10.10.11.9
          prefix-length: 24
        dhcp: false
        enabled: true
  capturedStates:
    address-from-traffic-1:
      ip: 10.10.10.8
      prefix-length: 24
    address-from-traffic-1:
      ip: 10.10.11.9
      prefix-length: 24
  ippool:
    traffic-1:
      allocated:
      - 10.10.10.8
    traffic-2:
      allocated:
      - 10.10.11.9
```

### API Extensions baremetal IPI

At baremetal IPI the Nmstate configuration is passed to openshift installer
using the baremetal host `networkConfig` field [1].

[1] https://github.com/openshift/installer/blob/ca8aacd3c931eefabfddafb49cb44b76f24653a6/docs/user/metal/install_ipi.md#describing-hosts

To accommodate the ippools, capture and nmstate new fields will be needed there

> the following is an example of the install-config.yaml from openshift with

Author: I am going to remove baremetal IPI and day 0 day 1 from here

content of the day 2 `NodeNetworkConfigurationPolicy` and also the `ipPools`
since there is not going to be multiple `networkPolicy` fields.

### Implementation Details/Notes/Constraints

#### Day 2

Since kubernetes-nmstate has a distributed architecture each of the handler
pods will calculate an allocatable ip from the `NodeIPPool` status and
retry the process on conflict

Handlers will check NNCEs at Reconcile and if the ip has being already assigned
it will re-use that and update `NodeIPPool` status with it, this way the
`NodeIPPool` can be reset but IPs are kept.

To optimize on storage and lookup instead of using a list of allocated offset
it's possible to use a bitmap, the OVN project use that approach to store the
offsets on a range [1], a snapshot of the bitmap can be stored as base64 at the
`NodeIPPool` status.

For the scale down scenario, the current kubernetes-nmstate architecture
contains a controller for node that will Reconcile on node deletion, but since
the nodes may be drained before deletion the pod running the controller
will no longer be running, this means that the controller may need to run at
control-plane nodes and watch for all the nodes.

> [1] https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/ipallocator/allocator.go

Reviewer: Does this section make sense here ? Like it seems to not fit - at all - between the `Baremetal IPI` and the `API Extensions baremetal IPI` ? It seems to be a simple general purpose implementation detail. 

It totally broke my concentration on anything related to baremetal.

Author: Going to rework the doc to remove baremetal IPI, that has to be done at a follow up.

#### Baremetal IPI

The `image-customize-controller` [1] is a golang controller so it would be quite
easy to import ippool library from kubernetes-nmstate and nmpolicy to render the
`networkPolicy`.

[1] https://github.com/openshift/image-customization-controller/blob/3ad18911d14dde1c23b660f99bb05e12a21c99e4/pkg/ignition/builder.go#L79

### Risks and Mitigations

If a bug prevents IPs deallocation the ippool can be exhausted and end up
with network reconfiguration or scaling up failures. To fix those situations
the field `status.allocated-offsets` at `NodeIPPool` can be set with an
empty list triggering a pool reconstruction, bringing back the
stale IPs. This risk gets higher for shorter IPPools.

If the `NodeIPPool` is exhausted the values from the `spec` fields can be
changed to add more free IPs to the pool. Doing so involves that handlers will

> reconstruct `status.allocated-offsets` with the already reserved IPs at NNCE.

Reviewer: didn't we say this is bad ?

### Drawbacks

Updating the pool is distributed across nodes by the nmstate-handler pod, this
means that N nodes will try to update `NodeIPPool` and retry on conflict. Is
not expected the IPAM mechanism will run very often so this should not be a

> problem.

Reviewer: famous last words :) 

I unfortunately don't see a way to mitigate this risk.

Debugging the allocated offset can be a challange since it involve a convertion

> from the offset to the IP is pointing too depending on `NodeIPPool` spec.

Reviewer: Well, as long as there's a translation mechanism involved, we can always create a tool to help out.

It would make it definitely simpler. Consider adding it to the implementation schedule (very end of the list).

Fixing the "exhausted" issue updating the NodeIPPool involves a lot of logic to
reconstruct the `allocated-offsets` that has to be carefully implemented and
tested, what alternate would be to just delete `NodeIPPool` and create new one,

> so `NodeIPPool` is a read-only resource.

Reviewer: I honestly wouldn't even bother. Plan well in advance, use huge ranges if you plan on having lots and lots of hosts. 

But your current proposal seems more conservative than the previous option. I'm OK with it.

The baremetal IPI Nmstate does an offline NetworkManager keyfile generation so
there is no network `currentState` that nmpolicy can reference to, so using
nmpolicy just to reference the ippools may be too overkill.

> If in the future nmstatectl is included at FCOS then the `networkPolicy` will

Reviewer: What's an FCOS ? (link to the fedora coreOS page please).

Author: Removing baremetal IPI reference.

- Expression "ippool.[name].allocate" feature either at kubernetes-nmstate or nmpolicy they have to call the created IPAM mechanism
- Integrate IPAM, nmpolicy and kubernetes-nmstate for simplest use case, allocate an address for the first time.
- Cover the deallocation scenario where the pool is unreferenced
- Cover the scale up scenario
- Cover the scale down scenario
- Cover restart scenarios
- Add `networkPolicy` field at baremetal API [openshift-installer API](https://github.com/openshift/installer/blob/master/pkg/types/baremetal/platform.go#L85)
- Glue `networkPolicy` field with IPAM + nmpolicy at [image-customization-controller](https://github.com/openshift/image-customization-controller)
- Test baremetal IPI
- Add automatic garbage collector for stale IPs

## Alternatives

An alternative is to implement directly an IPAM feature at Nmstate with a
plugin system for ippool retrieval and allocation storage so at day 2
it will use the `NodeIPPool` and `NodeNetworkConfigurationEnactment` and
at baremetal IPI It will read the ippool from the openshift installer config
and store the allocations at the nodes file system, later on it will be

> converted into proper NodeIPPool and NNCEs using different storage plugin.

Reviewer: I miss a comparison between the proposed solution and the alternative. 

Why did you choose one over the other ? What were its advantages over this one ?

Author: Well there are not much alternatives after looking around.

# Host IP pooling with NNCPs.

## Summary

Provided a mechanism to configure automatically static IP configuration for secondary interfaces on existing and future Nodes

## Motivation

Using DHCP as the IP assignment mechanism has some drawbacks:
- Introduce latencies on node restart.

> - Has problematic behavious when the DHCP server is miss configured.

Reviewer 6: Well, isn't everything problematic, when misconfigured?

```suggestion
- Has problematic behaviours when the DHCP server is misconfigured.
```

- It depends on network setup that can be complex.

> To overcome this static configured interfaces taking the address from a

Reviewer 6: ```suggestion
To overcome the above-mentioned drawbacks, statically configured interfaces taking the address from a
```

dynamic host IP pool is a solution. Since the interfaces are static there is no
reconfiguration after some time or at node restart.

This functionality was requested on U/S via https://github.com/nmstate/kubernetes-nmstate/issues/725

> Spike using whereabouts https://github.com/nmstate/kubernetes-nmstate/pull/1081

Reviewer 6: I understand that this effort is just a spike, or poc, but the solution there seems to be in conflict with
this comment from the U/S issue: https://github.com/nmstate/kubernetes-nmstate/issues/725#issuecomment-1021906027

Shouldn't the desing of making this knmstate feature agnostic to the used IPAM plugin be part of the proposal?

Author: I can remove the PoC reference, is just added to have a little of background, but maybe is not adding much.

Reviewer 6: I think the poc reference can remain here.

> The part that is going to be agnostic is the storage backend more than the IPAM itself.

This would be nice to mention somewhere, unless it already is in the proposal below that I'm just about to read.

Author: Keeping the spike and rewording it.

### User Stories

- As a cluster administrator, I want to configure static IPs on secondary NICs on a pool of nodes without creating one per each node.
- As a cluster administrator, I want scale up the cluster with new nodes getting IPs from the policy.
- As a cluster administrator, I want each of these static IPs to be unique, and taken from a defined range, and freed when unused.

### Goals

- Allow users to use NodeNetworkConfigurationPolicy to configure static cluster-unique IP addresses without hard-coding them.
- If at all possible, we should pick a solution that could be utilized even in the IPI nmstate API.

### Non-Goals

- This design does not cover IP address configuration of management NIC on day-0/1

## Proposal
## Implementation History

- Day 2 ippool with kubernetes-nmstate is implemented.
- The ippool library at kubernetes-nmstate is extracted or make it public.

> - Baremetal IPI is implemented adding `networkPolicy` field at install config.

Reviewer: I can't make any sense of this proposal with just this info.

Author: Re-worked it
